Fix multibyte character support for `highlight-indent-guides-character`.

hongyi-zhao opened the pull request:
Fix https://github.com/DarthFennec/highlight-indent-guides/issues/132

Using non-ASCII characters like `│` causes error: `"Attempt to store non-ASCII char into multibyte string"`.

The issue: `aset` doesn't work properly with multibyte characters when modifying strings created by `make-string`.

Solution: Build the string from a character list using `(apply 'string chars)` instead of modifying an existing string with `aset`.

Tested with both ASCII (`|`) and Unicode (`│`, `┃`, `┊`) characters.

hongyi-zhao committed:
**Fix multibyte character support for `highlight-indent-guides-character`**

Using non-ASCII characters like `│` causes error: `"Attempt to store non-ASCII char into multibyte string"`.

The issue: `aset` doesn't work properly with multibyte characters when modifying strings created by `make-string`.

Solution: Build the string from a character list using `(apply 'string chars)` instead of modifying an existing string with `aset`.

Tested with both ASCII (`|`) and Unicode (`│`, `┃`, `┊`) characters.

jcs090218 commented:
It works fine on my end. 🤔 Can you post the backtrace here? Thanks!

hongyi-zhao commented:
<img width="1324" height="1011" alt="image" src="https://github.com/user-attachments/assets/01c347fb-cf81-45a5-a279-59cbeba9794d" />

hongyi-zhao commented:
The core issue is demonstrated by this simulation of what the original code does:

```elisp
;; Original approach in highlight-indent-guides--guide-string:
(let* ((cwidth 10)
       (showstr (make-string cwidth ?\s)))  ; Creates unibyte string with spaces
  (aset showstr 0 ?│))  ; Attempts to store multibyte char → ERROR

;; In Emacs 31+: ERROR "Attempt to store non-byte value into unibyte string"
;; In Emacs <31: May work due to less strict multibyte/unibyte enforcement
```

<img width="923" height="751" alt="image" src="https://github.com/user-attachments/assets/1fe46271-748e-4996-8d88-008aa6da518e" />

## Emacs Version Dependency

- **Emacs 31+** (including current development version): Stricter multibyte string handling causes the error
- **Emacs 28-30**: May not show the error consistently
- **Earlier versions**: Generally more lenient with mixed byte/multibyte operations

## The Fix's Core Idea

Instead of creating a string and modifying it with `aset`, build the string from a list:

```elisp
;; New approach (works in ALL Emacs versions):
(let ((chars (make-list cwidth ?\s)))  ; Create list of characters
  (setcar chars ?│)                    ; Modify list (handles multibyte)
  (apply 'string chars))                ; Convert list to string
```

The key insight: Lists can hold any character type naturally, while `make-string` with ASCII creates unibyte strings that cannot be modified to contain multibyte characters via `aset` in modern Emacs.

This change ensures compatibility across all Emacs versions while properly supporting Unicode guide characters.

Testing environment: GNU Emacs 31.0.50 (build 2025-08-28)

Regards,
Zhao

mattiase commented:
(I'm here because of a question on the emacs-devel list.) This indeed needs to be fixed in order to work with Emacs 31. The patch is fine; alternatives mostly differ in style.

For example, I'd suggest using a vector instead of a list. Instead of duplicating the code to handle properties, consider gathering them in a list and apply them afterwards. Essentially,

`make-string` -> `make-vector`
`(add-text-properties BEG END PROP STR)` -> `(push (vector BEG END PROP STR) props)`

and then do something like
```
  (let ((showstr (concat char-vector))
    (dolist (p props)
      (add-text-properties (aref 0 p) (aref 1 p) (aref 2 p) (aref 3 p) showstr))
    ...)
```
well, you get the idea. But again, there is nothing fundamentally wrong with the proposed patch.

One thing though: I recommend not using the RESULT parameter in `dotimes` and `dolist`; it doesn't improve readability and may generate stray compiler warnings. Just move the result to after the loop.
Build with warnings as errors as far as possible; when warnings cannot be avoided, ask for advice.

hongyi-zhao committed:
Fix multibyte character support using vector approach

Use vector instead of list as suggested by @mattiase (Emacs core developer).
- make-vector + aset works correctly with any character type
- concat converts vector to string, handling multibyte properly
- Collect text properties and apply after string creation

Fixes #132
